rangeserver/src/prefetching_buffer.rs

```diff
@@ -0,0 +1,946 @@
+use bytes::Bytes;
+use std::collections::BTreeMap;
+use std::collections::BTreeSet;
+use std::collections::HashMap;
+use std::sync::Arc;
+use tokio::sync::{watch, Mutex};
+use uuid::Uuid;
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub enum KeyState {
+    Fetched,        // Key has been fetched and is in the BTree
+    Loading(u64),   // Key has been requested and is being fetched
+    Requested(u64), // Key has been requested but fetching has not been started
+}
+
+#[derive(Default, Debug)]
+pub struct PrefetchingBuffer {
+    pub prefetch_store: Arc<Mutex<BTreeMap<Bytes, Bytes>>>, // stores key / value
```

Comment:
I suspect there are way too many mutexes here. These are not free performance wise, but also complicate the code since you have to be careful to lock them in the right order to avoid deadlocks.

Is it possible to create a struct(s) to store the state you need on, and then wrap it in one/few mutexes instead?